File: AdventOfCode2022/Day04/Day04.hs
import Data.List (intersect, isSubsequenceOf)
import Data.List.Split (splitWhen)

rangeToList s = [left, left + 1 .. right]
  where
    [left, right] = map (\x -> read x :: Int) $ splitWhen (== '-') s

contains line = if isSubsequenceOf elf1 elf2 || isSubsequenceOf elf2 elf1 then 1 else 0
  where
    [elf1, elf2] = map rangeToList $ splitWhen (== ',') line

intersect' line = if not (null common) then 1 else 0
  where
    [elf1, elf2] = map rangeToList $ splitWhen (== ',') line
    common = elf1 `intersect` elf2

main = do
  content <- readFile "input.txt"
  print $ sum . map contains $ lines content
  print $ sum . map intersect' $ lines content

File: AdventOfCode2022/Day04/day04.js
const fs = require("fs");
const data = fs.readFileSync("Day04/input.txt", "utf8");

const lines = data.split("\n").map((x) => x.trim());

const overlap = [];

const overlap2 = [];

for (let i = 0; i < lines.length; i++) {
  const line = lines[i];
  const [elf1str, elf2str] = line.split(",");
  const [elf1left, elf1right] = elf1str.split("-").map((x) => +x);
  const [elf2left, elf2right] = elf2str.split("-").map((x) => +x);

  if (
    (elf1left <= elf2left && elf1right >= elf2right) ||
    (elf2left <= elf1left && elf2right >= elf1right)
  ) {
    overlap.push(line);
  }

  if (
    (elf1right >= elf2left && elf1left <= elf2right) ||
    (elf2right >= elf1left && elf2left <= elf1right)
  ) {
    overlap2.push(line);
  }
}

console.log(overlap.length);

console.log(overlap2.length);
